### DISCUSSION TITLE
Виджеты

### DISCUSSION THREAD

**Discussion by tipodanet** [k-ivan/visual-bookmarks-chrome#179] [Ideas]
Иван, приветствую. 
Хотите чужую идею ? В смысле обсудить с автором -- aminought
Он наш местный. Сделал шикарный виджет пол Chromium , просто суперский . 

https://github.com/aminought/vivaldi-widgets/blob/master/widgets.js

Хоть Вивальди из-за своей нулевой поддержки юзеров (по сути весь их форум -- варимся в собственном соку, разрабам до форума дела нет), но он имеет возможности для кастомизации CSS и JS. 
В этой части все открыт, код сами себе юзеры не шифруют :).
А вот внутренний код сами вивальдовцы обфусицируют (вроде так пишется). 
Автор сделал виджеты , все работает. Может, были еще какие мысли , углубить код. Но от разрабов реакции ноль.
Он плюнул и ушел на FF. Там весь код открыт. 

Чтобы было представление о чем речь -- это просто ссылка на часть другого сайта, что-то вроде <iframe>:
[vidget](https://i.imgur.com/Xw0wSt1.mp4)
Я уже не помню, по-моему это просто виджет со ссылкой на  что-то из  Three.js .
Простой тег <body> тоже прекрасно работает.

По размеры виджеты можно сделать одинаковыми с плитками , они могут отлично вписаться в сетку вашего расширения.
обсудите с автором, у меня предположение , что он приостановил разработку. 

Еще у меня 2 виджета попроще -- график курсов валют и погода

**Comment by k-ivan** [comment]
Здравствуйте!

Тут вариантов 2:

* делать коллекцию эффектов(можно с помощью сообщества в том числе) и потом в настройках давать возможность выбора какого-нибудь красивого эффекта на фоне(`canvas`, `webgl` без разницы, реализовывал такой вариант во втором своем расширении). Логика то или иного эффекта будет сразу присутствовать в кодовой базе, что решит проблему с трудностями внедрения пользовательских скриптов.
* Искать способ для добавления скрипта самим пользователем(но так чтобы это не очень проблемно)

По второй части курса и погоды, можно, но совершенно не понятно где и как это будет располагаться, так как сетка из плиток закрывает собой все пространство и некуда выносить такие штуки. Нужно думать, как можно и нужно-ли это.

**Reply by tipodanet** [reply]
Иван, согласен, что для нормальных людей сложно объяснить , что куда. 
Для продвинутых — объяснить принцип , указывается ОДИН селектор, и все. Нужно его найти самому, на своем сайте. Я делал свою простую страницу, без обвесни навигацией. Указал body, ✌️

**Reply by tipodanet** [reply]
У автора, конкретно под Вивальди, (у Вивальди собственная Экпресс-панель, такая же как ваши Визуальные.).  У него виджеты выглядят также , как и плитки, идут ниже плиток. 
Скорее всего, человек сделает себе 3-4 виджета и успокоится. Мне было бы удобно виджеты выше плиток, или сделать настройку

**Reply by k-ivan** [reply]
> Иван, согласен, что для нормальных людей сложно объяснить , что куда. Для продвинутых — объяснить принцип , указывается ОДИН селектор, и все. Нужно его найти самому, на своем сайте. Я делал свою простую страницу, без обвесни навигацией. Указал body, ✌️

не совсем понял, о чем речь идет

**Reply by k-ivan** [reply]
> У автора, конкретно под Вивальди, (у Вивальди собственная Экпресс-панель, такая же как ваши Визуальные.). У него виджеты выглядят также , как и плитки, идут ниже плиток. Скорее всего, человек сделает себе 3-4 виджета и успокоится. Мне было бы удобно виджеты выше плиток, или сделать настройку

можно скрин, как это выглядит?

**Reply by tipodanet** [reply]
виджету нужен только селектор, размеры он сам ставит, все одинаковые, все красиво становятся в сетку. Но виджеты идут отдельно , своим рядом, со своим общим размером. Юзер сам указывает этот размер, можно его сделать через flex, и все виджеты будут одинаковы с вашими плитками

**Reply by tipodanet** [reply]
надо включать Вивальди. Выглядит просто — такие же квадраты, с окошками  других сайтов. У меня был кусок курса валют с главной страницы EDGE. там красивые курсы

**Reply by tipodanet** [reply]
снаружи виджет даже подчиняется CSS. В видосе видно , что я ему красную рамку нарисовал.

**Reply by tipodanet** [reply]
знаете, где самая навороченная погода 🌡️ ? Это Ventusky. Круче погоды нет. В этом сайте я тоже нашел внутренний класс, без внешнего содержания. Р вся эта шикарная погода тоже становится внутрь виджета. 
Это просто мини сайт, виджет ничего не ограничивает. Такой виджет с погодой тоже поддерживает мышь, перемещение карты и переключение режимов погоды.

**Reply by k-ivan** [reply]
> виджету нужен только селектор, размеры он сам ставит, все одинаковые, все красиво становятся в сетку. Но виджеты идут отдельно , своим рядом, со своим общим размером. Юзер сам указывает этот размер, можно его сделать через flex, и все виджеты будут одинаковы с вашими плитками

понятно, нужно думать над дизайном, или искать макет прототип

**Reply by k-ivan** [reply]
> надо включать Вивальди. Выглядит просто — такие же квадраты, с окошками других сайтов. У меня был кусок курса валют с главной страницы EDGE. там красивые курсы

у меня нет Вивальди )

**Reply by k-ivan** [reply]
> снаружи виджет даже подчиняется CSS. В видосе видно , что я ему красную рамку нарисовал.

уточните, в каком видео?

**Reply by k-ivan** [reply]
> Ventusky

там нет готовых виджетов для встраивания, типа следите за новостями(так написано на сайте), но это просто к слову, все равно не понятно как бы оно выглядело, если бы было

**Reply by tipodanet** [reply]
"уточните, в каком видео?" — ссылка в 1 посте.
"там нет готовых виджетов для встраивания, " 
Сначала. 
На гитхабе есть автор, с адским ником Aug......
Он сделал виджеты, которые встраиваются на главную страницу Вивальди. Принимают внутрь ЛЮБЫЕ сайты. Можно даже указать адрес сайта и body, для любого сайта. И получаеМ виджет С окошкоМ этого сайта. Представить не сложно.
Селектор нужен, Чтобы ограничить Нужную область сайта.

**Comment by tipodanet** [comment]
к вопросу надо или нет, уже есть готовые расширения [Casca](https://chromewebstore.google.com/detail/nifaiikfnpnppeicpgannigjabpjdcmh) довольно известное уже. И там тоже есть отдельные виджеты, и там как раз огромный выбор. Там большой список, но там  все виджеты стандартные.  Юзер сам не может поставить свой адрес. Это резко снижает его ценность для меня. Там тоже есть : фотки с фейсбука случайные. И погода там навалом разных вариантов. И все они являются виджетами на странице браузера. И тоже поддерживает перемещение, вроде размер можно менять. Но своего селектора там не укажешь , в этом огромная разница. Я поиграл с теми виджетами и удалил их. Мне мусор не нужен, мне нужны мои виджеты. Так что эта идея не новая, и она уже активно развивается.

**Reply by k-ivan** [reply]
Не спроста там то, что только сделали сами разработчики, я повторюсь очень проблемно сейчас вставлять сторонний код в свое расширение, очень строгие политики и они становятся все строже.

**Reply by tipodanet** [reply]
Да, нужен любой адрес, и селектор. Готовые виджеты бессмысленны. И они уже есть

**Reply by tipodanet** [reply]
А разве нельзя Как-то ограничить области действия JS ? Чтобы скрипты только отображали страницу, и не лезли в пароли браузера ?

**Reply by tipodanet** [reply]
Я понимаю, что можно создать страницу специально для кражи паролей и через виджет залезть в эти пароли. Ну вот, было бы хорошо, чтобы виджет мог ПоказываТь только визуальную часть, а, например, все формы для ввода Текстовой информации Не работали. Было бы хорошо.

**Reply by k-ivan** [reply]
это так не работает, это работает так, что политика строга к любому коду, который встраивается из стороннего источника

а уже что там будет, картинка котенка или код, который что-то тырит дело вторичное

**Reply by tipodanet** [reply]
Код виДЖета -- это часть приложения, он нИчего Не кРадет и Ничего не Встраивает. виджет принимает только Адрес сайта. 
А в чем разница видеть Адрес сайта через видЖет или Через страницу браузера ?

**Reply by tipodanet** [reply]
Защитой От хакеров занимается сам браузер, и виджет ее никак не обходит. Там же код есть 20кб, можно все смотреть.

**Reply by k-ivan** [reply]
защита браузера состоит только в том, чтобы не давать Вам запускать посторонний код, и по большей части заключается это только в предупреждениях, браузер никакой защитой от хакеров не занимается!!!

**Reply by tipodanet** [reply]
Это лучшая новость .
Значит, основной проблемой для развития вашего расширения, Иван, Являетесь вы сами :)
Откройте глаза. Из 30 000 юзеров Вашего расширения  предположу что 1000 залезли в разрешЕния , остальные 29 000 -- это баБушки , блогерши, алкоголики и другие бездельники :)
Вы немного переоцениваете аудиторию интернета.
Мне кажется, вам гораздо интереснее иметь 1 000 000 алкашей, чем 30 000 программистов

**Reply by k-ivan** [reply]
очень коротко и поверхностно, если все еще будут вопросы, Вы будите знать, про что почитать

Например речь о встраивании своего кода в расширение. Допустим мы сохранили код как строку, затем нам нужно его выполнить на странице нашего расширения, в новом манифесте мы не сможем его выполнить, так как строку нужно будет пропустит через спец. функции (eval или new Function), а это запрещено браузером в манифесте v3. 
У нас есть апи chrome.scripting.executeScript и оно тоже нам не поможет в данном случае, так как не сработает на странице расширения и предназначен для встраивания на сайты.
А еще есть chrome.userScripts и эта штука требует от пользователя включения режима разработчика в браузере и по другому просто не работает

К чему я, да к тому, что это все даже работать нормально не будет в рамках расширения, так как если Вы запросите host_permission <все_хосты> код будет запускаться на всех сайтах. Нет на данный момент хорошего решения встраивания ваших скриптов на страницу моего расширения.

Это очень коротко о ситуации вокруг доступного апи, никто не боится гугла или другого вендора, просто нет возможности(я по крайней мере ее не вижу на текущий момент), чтобы встраивать ваши скрипты в мое расширение.

Скрипты, которые есть в рамках самого расширения выполнять проще, чем те, что приходят из стороннего источника в режиме реального времени.

**Reply by k-ivan** [reply]
Если бы это нормально работало, давно бы добавил, мне не жалко

**Reply by tipodanet** [reply]
Мы с вами  как на разных языках. Лучше бы на HTML, как то понятнее. 

`Скрипты, которые есть в рамках самого расширения выполнять проще, чем те, что приходят из стороннего источника в режиме реального времени.`

Я могу это отнести а старому вопросу о доступе к JS  для юзера, такому же, как сейчас есть для CSS.
Ну нет, так нет. Мне он особо и не нужен был, там могли только .foreach() пригодиться. 
А сегоднЯ вы сказали , что оказывается Гугл не запрещает, А только предупреждает.
А разве в Расширениях типа Tempermonkey нет своих JS ?

Но я о другом. Где В виждетах с открытым кодом Доступ юзера к JS, у юзера даже CSS нет.
У него только адрес, селектор, размер, zoom. Это все.
Скопируйте код виджета в расширение и все. 
Амиго сделаЛ форму Для ввода перечисленного.
Никаких целых страниц для Опасных скриптов.
6 полей для 6 строк.
Я уже 100 раз сказал, что это просто ссылка на сайт. Мелкое Окно в браузере на другой сайт.
Там нет места юзерскому JS, CSS.
Щас сниму видео.

**Reply by tipodanet** [reply]
Открытый JS -- Это только в Вивальди. Он Допускает Модифицировать window.html . Открыть,  подключить любой скрипт. И перезапустить браузер и скрип доступен. А судя по тому, что Вивальди- -- это копия хрома. Надо мне И с хромом то же самое попробовать. Вдруг прокатит,  где-нибудь появятся эти виджеты.

**Reply by tipodanet** [reply]
Если Гугл будет проверять расширение с Виждетами, он увидит только для юзера 6 Полей. JS нет, CSS тот Же, что и к остальным плиткам

**Reply by k-ivan** [reply]
Как ссылка ваша создаст ваш виджет? Можете объяснить, по ссылке должен быть  js код, который что-то будет делать, вот этот самый js код надо запустить в рамках расширения. Либо эта ссылка должна встраиваться в iframe на странице. По другому ваша ссылка не создаст магическим образом какой-то виджет.

**Reply by k-ivan** [reply]
Учитывая, что iframe отпал из комментариев выше, значит по ссылке js, и значит его надо выполнять в рамках расширения, а дальше я уже писал

**Reply by tipodanet** [reply]
я уже копировал вам код, который доступен юзеру . 
Вот еще  входные  юзерские данные для виджета курсов валют
```
            id: 'CurrencyEdgeWidget',
            url: 'https://www.msn.com/ru-xl/money/chart?id=avyn9c&timeFrame=10Y&chartType=area',
            zoomFactor: 1,
            selector: 'div.chartContainer-DS-EntryPoint1-1',
            width: '612px',
            height: '408px',
            timeout: 3000
```

<img src="https://i.imgur.com/aKz1UtS.jpeg">

Этот URL   я сам нашел на странице msn.com. 
Здесь нет JS, нет запретов.
А код из 1 поста, рисует окошко в браузере. Больше ничего.  Дальше работает сам браузер. И рисует тоже самое , что и в новой вкладке. Только ограниченное  селектором и Zoom-facktor .

**Reply by k-ivan** [reply]
То что вы скинули, все это вставляется с помощью фичи, под названием WebView, то есть это какая-то вивальдивская тема, и работает она только там.

**Reply by tipodanet** [reply]
все правильно. Этот виджет использует классы Вивальди и вставляется  в его Expess-Panel. 
Перепишите  код под свои классы -- будет вставляться в ваше расширние.

**Reply by k-ivan** [reply]
Какие свои классы, нет в хроме api такой сущности как WebView, через который вы видите сайт в мини окне Вивальди.

**Reply by tipodanet** [reply]
WebView всплывает только в мобильных браузерах. И то как не основной . Основые браузеры под Андроид на мобильном Хроме. Включая Вивальди. Некоторые на WebView,  типа SOUL , 
Мы на ПК,  причем здесь WebView ?

**Reply by k-ivan** [reply]
При том что в вашем Вивальди есть такое апи

**Reply by tipodanet** [reply]
Amigo ушел с Вивальди на FF . Там тоже есть WebView , которого нет в Chrome  ?

**Reply by k-ivan** [reply]
Я не знаю кто такой Амиго, и понятия не имею есть в FF этот апи или нет, но это штука точно есть в Вивальди и вокруг нее крутится код по вашей ссылке.

**Reply by tipodanet** [reply]
**WebView не используется в браузере Vivaldi для Windows. Vivaldi для Windows, как и Google Chrome для Windows, использует свой собственный движок рендеринга, основанный на Blink, который является форком WebKit.

Vivaldi - это независимый браузер, разработанный компанией Vivaldi Technologies. Он не использует компоненты Android, такие как WebView, поскольку Vivaldi предназначен для работы на настольных операционных системах, таких как Windows, а не на мобильных платформах.

WebView является специфичным для Android и не применяется в настольных браузерах, включая Vivaldi для Windows.**

**Reply by tipodanet** [reply]
Имя  ### aminought  есть в ссылке в 1 посте. Я думаю , его бы никакой WebView  в Chrome не остановил, особенно когда его там нет. 
Даже мне видно в коде, что там больше про CSS через JS.

**Reply by k-ivan** [reply]
> 
> **WebView не используется в браузере Vivaldi для Windows. Vivaldi для Windows, как и Google Chrome для Windows, использует свой собственный движок рендеринга, основанный на Blink, который является форком WebKit.
> 
> Vivaldi - это независимый браузер, разработанный компанией Vivaldi Technologies. Он не использует компоненты Android, такие как WebView, поскольку Vivaldi предназначен для работы на настольных операционных системах, таких как Windows, а не на мобильных платформах.
> 
> WebView является специфичным для Android и не применяется в настольных браузерах, включая Vivaldi для Windows.**
> 

Блин, ну посмотрите код, который мне сами же скинули, для чего эти тексты из Википедии. В вашем коде используется инстанс WebView с кучей методов. Не верите, смотрите код по ссылке.

**Reply by k-ivan** [reply]
> Имя  ### aminought  есть в ссылке в 1 посте. Я думаю , его бы никакой WebView  в Chrome не остановил, особенно когда его там нет. 
> Даже мне видно в коде, что там больше про CSS через JS. 

Раз вы разбираетесь лучше, флаг в руки.

**Reply by tipodanet** [reply]
как я колупал когда то CSS Vivaldi , видел него классы. И только на 883 строке я нашел хоть что-то ,  что напоминает о Vivaldi 
```
   get #speedDial() {
            return this.#internalPage?.querySelector('.dials.speeddial');
        }

```
все остальное  -- это чистый JS. Никакого отношения не имеет к каким либо сложным VebView, другим скрытым технологиям. 
Все переменные и константы   aminought   создает сам , тусует селекторы туда-сюда. 
Его код будет работать везде,  где есть  чистый JS.

**Reply by k-ivan** [reply]
Нет, помимо js там есть внутренний веб-компонент webview со своими методами и этот компонент откуда то взят, скорее всего предоставлен Вивальди и все крутится вокруг него.

Я больше не буду ничего доказывать, Вы не видите очевидных вещей или не хотите.

**Reply by tipodanet** [reply]
от разрабов вивальди вы никогда ничего не дождетесь. Это известно. Они шифруют свой код. И если у них попросить свернуть все папки, юзеры будут ждать этого 6 лет. 

```
  #createWidgetSaveButtonListener(saveButton, widgetWrapper, widgetInfo) {
            const widget = widgetWrapper.querySelector('.Widget');
            const webview = widgetWrapper.querySelector('webview');
```
Amino сам создает себе константы .

**Reply by k-ivan** [reply]
Константы тут вообще не причем и не главное, вся магия происходит внутри кастомного тега webview,  и все взаимодействие через его методы происходит. Стили, поиск по селектором эта фигня, вся суть в этом теге и его апи.

**Comment by tipodanet** [comment]
У авТора Виджеты встраивались на всех Вкладках Экпресс-Страницы ,  Ниже всех, одинаковые.
Но тут есть несколько вариантов:
-- выше
-- ниже
-- на всех
-- на главной
-- на отдельной

Это все можно отдать самому юзеру, через настройкИ.

**Reply by tipodanet** [reply]
Иван, свяжитесь с автором Амиго, может, чего решите.
Ему УЖе предлагали  Оформить все в отдельное Расширение длЯ виджетов. 
Но думаю, эта КоллаборациЯ вам выгодна

**Reply by k-ivan** [reply]
не, мне не так это интересно, чтобы я связывался по такому вопросу

**Reply by k-ivan** [reply]
я допускаю, что возможно что-то из подобного добавлю, но обещать я не могу, все это идет от желания и времени всем заниматься.

**Reply by k-ivan** [reply]
учитывая, что данные вещи не имеют отношения к работе моего расширения

**Reply by tipodanet** [reply]
Глупо ОтказыаптьСя оТ готового на 99%
Вы рассуждаете Так, как-будто залезли в голову всем юзерам

**Reply by k-ivan** [reply]
Бизнес-логика расширения - !это работа с закладками в визуальном представлении!, не анимации на новой вкладке, не большой взрыв из виджетов на главной. И конкретно в этом расширении маятник всегда будет в сторону представления закладок(весь мир крутится вокруг этой миссии, повторюсь, конкретно в этом расширении)

Я допускаю, что можно добавить например канвас на задний план или webgl(и у меня было это в планах), я допускаю, что можно подумать и над виджетами(но основательно, дизайн, как что, куда и т.д. и т.п.). 

И готового ничего нет, я буду писать код под свое расширение, мне ничем не помогут ссылки выше.

**Reply by k-ivan** [reply]
Судя по Вашим комментариям, мне кажется, Вас вообще не интересует работа с закладками и это нормально, но возможно Вам нужны расширения типа momentum, которые как раз показывают Вам красивую картинку при входе и кучу виджетов, то есть есть много расширений у которых суть именно в этом заключается.

**Reply by tipodanet** [reply]
Вам НУЖНА (пишу большими) , а еще точнее ВАМ любая изминка, чтобЫ отличаться от других.
Визуальных закладок - с десяток наБерется в Сторе.
Все отличие только  В Дизайне. СлабовАто основа для успеха.
Вам надо конкурировать С такими монстрами как FVD.
WebGL В фоне ? А код где Возьмете ? На том Же Three.Js ?  И вы говорите, что вас взрывы из грАфики не интересуют ?

Я не говорил о взрывах.
У меня было 3 виджета -- курс валюты, погода и мой личный сайт. 
У самого  Амиго был виджет с его статусом на этом гитхабе, больше не показывал.
У ваС есть ННастройка  стилей для своих плиток. Значит, вы рассчитываете на грамотных юзеров.
Дайте людям эти пустые квадраты на отдельной странице.  На обсуждение дизайна их рАМки ...  Это  не тема . Виджеты -- это квадраты, и Они прекрасно принимают тот Же общий css.

**Reply by tipodanet** [reply]
Это вам нужно отличаться от Других, а нЕ мне.
С бизнесОм понятно. 
Ну напишите свои виджеты , готовый код выложен.
Тем более, даже Амиго был не первый

**Reply by k-ivan** [reply]
Еще раз нет никакого готового кода, код нужно будет писать свой(в подробности вдаваться не буду). Напишу, когда и если будет время и желание.

> WebGL В фоне ? А код где Возьмете ? На том Же Three.Js

Что значит где возьму, а что нельзя написать свой код? Или взять реализацию адекватных эффектов с того же гитхаба.

> У самого Амиго был виджет с его статусом на этом гитхабе, больше не показывал.

мне не интересно, что там было в Амиго

> Визуальных закладок - с десяток наБерется в Сторе.

так и отлично, но по моему опыту, мои самые адекватные(именно по моему опыту, я пробовал их все из популярных)

> Это вам нужно отличаться от Других, а нЕ мне.

мне это не нужно

> Все отличие только В Дизайне. СлабовАто основа для успеха.

а каком успехе речь, я же писал расширение изначально для себя, для меня уже успех, что ~30k пользователей тоже пользуются

**Reply by tipodanet** [reply]
Еле нАшел momentum, их соТнИ (метро, такси). Этот momentum как Раз и является готовыми виджетами с чужими картинками. Задачи, будильники , не интересует.  Я найду Название тех стандартных виджетов, для примера -- CASCA. Там больше выбор , больше свалки. Это не тО.
Вы меня обвиняете, что мне не нравятся закладки. Нравятся, как без них. Мне нужен дополнительный функционал. 
Если  подумать , то следующим шагоМ развития закладок, то именно виджеты и являются их продолжением. 
В идеАле, если бы  у всех на компах стояли карты 4090, то можно было все закладки Делать в Виде гоТовых ЖиВыХ сайтов. 
Почему неТ ? Сразу вИдно Обновление, есть смысл ЗАХодить иЛИ нет.
Кстати, в Вивальди с иХ личным сервероМ,  Preview  обновляется само, и некоторые изменения Видны, это удобно. На картинке даЖе видно, чТо мне пришло ЛС. 
Т. Е  здесь Вивальди с его  визуальными круче.
Вам нужна любая идея, которая отличит вас от других.
И виджет не может поставить безопасность поД угрозу -- это просто ссылка на любой сайт, такая же как и на соседней вкладке.
Вы волнуетесь о безопасности больше , чем сам гугл.

**Reply by k-ivan** [reply]
> Вы меня обвиняете, что мне не нравятся закладки

Да где Вы увидели, что я Вас обвинял в чем-то, ну не было же такого. Я предположил, что Вам просто нужна другая категория расширений и судя по описаниям так и есть.

> В идеАле, если у всех на компах стояли карты 4090, то можно было все закладки Дать в Виде гоТовых
ЖиВыХ сайтов.

не, это очень плохо, ведь в закладках могут быть сотни сайтов, представляйте себе, кол-во запросов при каждом открытии вкладки.

> Кстати, в Вивальди с иХ личным сервероМ, Preview от обновляется само, и некоторые изменения Видны, это удобно. на картинке даЖе видно, чТо мне пришло ЛС.

такое можно провернуть в теории, правда есть много нюансов, и не очень тривиально

> Т. Е здесь Вивальди с его визуальными круче.

да я рад за них, только я один а их много, и денег у них явно больше, очевидные вроде вещи говорю, странно нас тут сравнивать 

> И виджет не может поставить безопасность поД угрозу -- это просто ссылка на любой сайт, такая же как и на соседней вкладке. Вы волнуетесь о безопасности больше , чем сам гугл.

Вам надо самому окунуться в разработку расширений, Вы просто не понимайте, какие есть ограничения для расширений, то есть даже вставить код от пользователя, это танцы с бубном из-за политик браузера в отношении расширений.

**Reply by tipodanet** [reply]
Нет, ссылка на сайт и голый селектор — это не код. 
Ссылка - это адрес, как и в адресной строке. 
Внутрь GetquerySelector(" ... ") опасный код сложно вставить. Так в гугл и напишите.
Я так понял, весь код виджета -- он только про размеры, и расположение рядом в сетку

**Reply by k-ivan** [reply]
> Нет, ссылка на сайт и голый селектор — это не код.

Я вообще не понимаю, о чем тут речь, вроде нигде про это я не писал

> Я так понял, весь код виджета -- он только про размеры, и расположение рядом в сетку

Да почему, внутри виджет может быть все что угодно и куча логики и много чего. То про что Вы говорите похоже на Iframe.

**Reply by tipodanet** [reply]
Вообще ту ссылку дал для вас, а не для себя. В меня этот код есть и работает. 
Вы ее не открывали.
```
const WIDGETS = [
        {
            id: 'VivaldiProfileWidget',
            url: 'https://forum.vivaldi.net/user/aminought',
            selector: '.profile.row',
            zoomFactor: 0.8,
            width: '292px',
            height: '266px',
            timeout: 0
        }
    ];
```
Больше юзеру не нужно.
Вы ожидаете запрета 🚫 от Гугл этого кода ? Если пробьетесь через секретаршу, то разрешат.

**Reply by tipodanet** [reply]
Я знаю, что iframe старая система, и гугл их не любит. Не индексирует в поиске. Нам индексация не нужна.
Иван, сделайте через iframe. Вероятно, Амиго переписал действие этого тега своими силами. 
Если вы сделаете сетку из iframe с ограниченим Области по селектору, это то Что надо.

**Reply by k-ivan** [reply]
еще раз, этот код вообще ничего мне не дает, Вы правда думайте, что я буду анализировать 1000 строк чужого кода?


проблема не в этом, проблема в том, что нужно запустить код из стороннего ресурса внутри расширения, а там много ограничений в политике безопасности, и скорее всего придется какие-то разрешения добавлять(это точно не мой путь)

**Reply by tipodanet** [reply]
А вы может с позиции разрешений объяснить , почему во вкладке я могу запустить любой сайт, а в вашем расширении - нет ?

**Reply by tipodanet** [reply]
В ваших плитках и в виждете даже общее поле есть : URL . представьте, мы  спорим и боимся Гугла, ☄️  из-за 1 строки — Selector: "... " . 
Одна строка !
А внутренности в квадрате отображает сам браузер, а не ваш код

**Reply by k-ivan** [reply]
> А вы может с позиции разрешений объяснить , почему во вкладке я могу запустить любой сайт, а в вашем расширении - нет ?

Вы можете почитать документацию по расширениям, раздел permissions, там много подробностей найдете.

**Reply by k-ivan** [reply]
я не боюсь гугла, я не хочу запрашивать у пользователя кучу разрешений

**Reply by tipodanet** [reply]
Т.е  Гугл вам разрешает , а вы брать Разрешения не хотите ? 
99, 9% ставят расширения и не лезут в их разрешения. Я  вхожу  в это число.
Немного оффтопа: есть сайт, там живой админ, легкое знакомство с ним. 
Я ема говорю, сделай темную/светлую тему. Я нахожу сайты, мНого,, не самые левые (например,  DuckDuckgo ), где вообще нет светлой темы, dark only.
В css Это делается 1 строкой (2 темы).
Я ему показываю  Разные опросы, периодически их мОжНо разных  найти , где 85% За темную тему.
Админ отвечает,  что ему 2 юзера сказали (Полноценный опрос он сам не проводил, у него всех юзеров на Сайте человек 50 Со мной), что у них нет доверия к Сайту с темным фоном, он подозрительный и прочая чушь.

Иван, вы согласны, что это проблема не большинства юзеров, а с головой у того админа ?
А виноват его сайт, и те 2 юзера, на ответы которых он наткнулся.
Вы точно уверены, что абсолютному подавляющему большинству юзеров есть дело до разрешений ? Мне нет . Я даже не читаю страницу настроек

**Reply by k-ivan** [reply]
> Вы точно уверены, что абсолютному подавляющему большинству юзеров есть дело до разрешений ? Мне нет . Я даже не читаю страницу настроек

И это очень плохо, не нужно пропагандировать это, как норму.

**Reply by tipodanet** [reply]
Хорошо, так просто iframe Требует разрешений, гугл Не против ? Гугл же Сам  и участвовал в его разработКе. там же все эти конторы типа W3C все завязаны.

**Reply by tipodanet** [reply]
Я предположу, что iframe не может выделить конкретный селектор из сайта.

**Reply by k-ivan** [reply]
Не обижайтесь пожалуйста, но у Вас нет понимания контекста всей сути, очень много хаоса, по типу, кто какие теги разрабатывал и прочие, хотя это не имеет никакого отношения к вопросу. Вам просто нужно попробовать самому сделать расширение, что погрузиться в суть детально.

**Reply by tipodanet** [reply]
Если iframe селектор не выделит, то не тратьте на него время

**Comment by aminought** [comment]
Привет, я автор этого мода для Vivaldi. Случайно увидел обсуждение, ну и дискуссию вы тут устроили. Сразу оговорюсь, читал по диагонали.

Не увидел нормальной ссылки, чтобы понять, что это вообще такое: [пост на форуме Vivaldi](https://forum.vivaldi.net/topic/95426/%D0%B2%D0%B8%D0%B4%D0%B6%D0%B5%D1%82%D1%8B-%D1%8D%D0%BA%D1%81%D0%BF%D1%80%D0%B5%D1%81%D1%81-%D0%BF%D0%B0%D0%BD%D0%B5%D0%BB%D0%B8), тут и демонстрация нормальная есть. WebGL и Three.js тут вообще ни при чем, мои виджеты — это просто оболочка для куска страницы любого сайта, и эту идею я позаимствовал из Arc Browser, там эта фича гораздо круче реализована, что-то вроде живых скриншотов, при этом нужную часть страницы можно выделить мышкой.

Как выше подметил @k-ivan, я для этого пользовался браузерным компонентом `<webview>`, к которому нет доступа из WebExtension API. Вообще, я далеко не веб-разработчик, но вроде бы можно положить страницу сайта в `<iframe>`, [достать](https://stackoverflow.com/questions/1088544/get-element-from-within-an-iframe) из него документ и поудалять всё, кроме указанного селектора, но многие сайты вообще не дают помещать себя в `<iframe>`. Да, мой код выглядит довольно простым, но лишь потому, что я использую встроенный компонент, который и делает всю работу.

В общем, я считаю, что автору этого расширения нецелесообразно внедрять подобную фичу, тем более она довольна нишевая.

**Reply by k-ivan** [reply]
Здравствуйте! Спасибо за развернутый ответ.

**Reply by tipodanet** [reply]
@aminought, привет. А я посмотрел по низкой активности вашего аккаунта. Здесь, на гитхабе, даже поискал на формах ФФокса. Что вроде бы вы не продолжали этих виджетов На фаер фоксе . Ну , решил, что вы занялись  другой деятельностью. 

Интересно, что вы оба разработчики и вы оба считаете, что это вещи нишевые. И никому не надо. 

Напомню Ивану, что надо как-то иногда взглянуть на свои разработки со стороны. Очевидно, что время реакции пользователя с  визуальными закладками примерно 1 секунда. У  меня занимает именно столько, я нажимаю  + В браузере,  открываются визуальные закладки, и мне хватает 1 секунды выбратт самую популярную , и на этом всё. Столько внимания я уделяю вашим закладкам , Иван, И я думаю , такая же ситуация и у других пользователей . 1 секунда, Поэтому вам необходима любая идея, чтобы удержать внимание пользователя. И виджеты к этому прекрасно подходят,  просто идеально.
 Вы считаете, что это лишнее, что 1 секунды внимания пользователя достаточно. aminought , как не странно, считает так же ;) 
Нишевая ? Только для грамотных ? У вас же есть поле для CSS, Найти селектор требует таких же знаний.

Здесь дело не в разработке, а о вашем личном отношении к этой идее, что вы оба не верите, что она кому-то, может быть, интересна . 
Довольно странная ситуация. 
Проведите опрос , соберите 1000 мнений, люди сами скажут, нужно это им или нет.

**Reply by tipodanet** [reply]
Не могу так уверенно спорить, но GPT прекрасно выдает все методы этого webview>, никая это не тайна :)

**Reply by k-ivan** [reply]
Вообще не понятно, какие методы и причем здесь GPT?

**Reply by tipodanet** [reply]
Доказательством того, что в визуальных закладках трудно что-то придумать новое, являются ваши конкуренты , Иван. Есть довольно известные закладки,  про них статьи написаны в интернете. Ну,  как альтернатива встроенным закладкам в браузер.

Так они написаны ещё на табличной вёрстке. Прикиньте, сколько им лет и сколько лет назад они уже остановились в развитии, если там вся вёрстка на   tr>. Вот с тех пор они и стоят на месте. А куда двигаться ? Квадратные зкладки , они у всех примерно одинаковые.

**Reply by k-ivan** [reply]
Вы что хотите конкретно, я уже Вам сказал, что нет такого апи ни в хроме ни в фф. 

Я же не спорю, с тем, что можно и нужно улучшать расширение.

**Reply by tipodanet** [reply]
@aminought, по крайней мере в >iframe> работает сложная фото-галерея 📸 на jQuery с анимацией, плеер музыкальный тоже. 
 Но >iframe> — это же на сайт положить ? Если выделить селектор, это тоже интересно, для любителей писать сайты
